crates/ra_hir_def/src/dyn_map.rs

```diff
@@ -0,0 +1,89 @@
+//! This is a work of fiction. Any similarities to Kotlin's `BindingContext` are
+//! purely a coincidence.
+
+use std::{
+    hash::Hash,
+    marker::PhantomData,
+    ops::{Index, IndexMut},
+};
+
+use anymap::Map;
+use rustc_hash::FxHashMap;
+
+pub struct Key<K, V, P = (K, V)> {
+    _phantom: PhantomData<(K, V, P)>,
+}
+
+impl<K, V, P> Key<K, V, P> {
+    #[allow(unused)]
+    pub(crate) const fn new() -> Key<K, V, P> {
+        Key { _phantom: PhantomData }
+    }
+}
+
+impl<K, V, P> Copy for Key<K, V, P> {}
+
+impl<K, V, P> Clone for Key<K, V, P> {
+    fn clone(&self) -> Key<K, V, P> {
+        *self
+    }
+}
```

Comment:
does derive not work for these impls?

Reply:
It won't for `Copy`, because `K` and `V` are generally not `Copy`

Reply:
Aah right